monix-execution/shared/src/main/scala/monix/execution/misc/Local.scala

```diff
@@ -60,12 +64,31 @@ object Local {
   def clearContext(): Unit =
     localContext.set(defaultContext())
 
-  /** Execute a block of code without propagating any `Local.Context`
+  /** Execute a synchronous block of code without propagating any `Local.Context`
     * changes outside.
     */
   def isolate[R](f: => R): R =
     macro Macros.isolate
 
+  /** Execute an asynchronous block of code without propagating any `Local.Context`
+    * changes outside.
+    */
+  def isolate[R](f: => Future[R]): Future[R] = {
```

Comment:
I am concerned about this overload.

The old one is a macro, this one is not. Also I know that the JVM cannot differentiate between `R` and `Future[R]`, especially here because we have a by-name parameter which translates into a `Function0`.

I'm thinking that we should no longer use a macro here. I'm not convinced that we gain anything by it. Macros are fine for internal functions, were we don't want the overhead of that closure, but for user code it just makes things complicated.

So I'm thinking that maybe we should introduce a type class, because why not:

```scala
trait CanIsolate[R] {
  def isolate(f: => R): R
}

//...

object Local {
  def isolate[R : CanIsolate](f: => R): R =
    CanIsolate[R].isolate(f)
}
```

Given that what we had was a macro, I don't think this would violate any binary compatibility.

Not sure about how this should behave in the context of `Task` though. Or if it's a good idea. Just a suggestion.

Reply:
I agree with macro, I meant to rewrite it to macro later but only because of consistency. I think we could safely remove it (and possibly reintroduce if it turns out it's needed). I know Oleg kept it for similar reasons

> So I'm thinking that maybe we should introduce a type class, because why not:
> ...

Hmm, I think it could work, we could have an instance for `Future` and then a lower priority instance for everything else. I can't tell if it's an improvement or not, to me it seems similar to having overload

> Not sure about how this should behave in the context of Task though. Or if it's a good idea. Just a suggestion.

`Task` has a different function for isolate on `TaskLocal` which does `bracket`. We could add an instance for normal `Local` in `Task` object I guess if it will be higher priority than default one for `Local` (not sure, needs to check)

WDYT @oleg-py ?

Reply:
I have no preference here. Typeclasses will work as well as `asInstanceOf`, really, unless you push it to properly isolate `Option[Future[Option[Future[A]]]]`. You'll also have to make sure it works properly with our `CancelableFuture`, that it doesn't fall back to lower priority instance.

> maybe we should introduce a type class, because why not

🤣